irc.go

```diff
@@ -0,0 +1,134 @@
+package twitch
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+type ircMessage struct {
+	Raw     string
+	Tags    map[string]string
+	Source  ircMessageSource
+	Command string
+	Params  []string
+}
+
+type ircMessageSource struct {
+	Nickname string
+	Username string
+	Host     string
+}
+
+func parseIRCMessage(line string) (*ircMessage, error) {
+	message := ircMessage{
+		Raw:    line,
+		Tags:   make(map[string]string),
+		Params: []string{},
+	}
+
+	split := strings.Split(line, " ")
+	index := 0
+
+	if strings.HasPrefix(split[index], "@") {
+		message.Tags = parseIRCTags(split[index])
+		index++
+	}
+
+	if index >= len(split) {
+		return &message, fmt.Errorf("parseIRCMessage: partial message")
+	}
+
+	if strings.HasPrefix(split[index], ":") {
+		message.Source = *parseIRCMessageSource(split[index])
+		index++
+	}
+
+	if index >= len(split) {
+		return &message, fmt.Errorf("parseIRCMessage: no command")
+	}
+
+	message.Command = split[index]
+	index++
+
+	if index >= len(split) {
+		return &message, nil
+	}
+
+	var params []string
+	for i, v := range split[index:] {
+		if strings.HasPrefix(v, ":") {
+			v = strings.Join(split[index+i:], " ")
+			v = strings.TrimPrefix(v, ":")
+			params = append(params, v)
+			break
+		}
+
+		params = append(params, v)
+	}
+
+	message.Params = params
+
+	return &message, nil
+}
+
+func parseIRCTags(rawTags string) map[string]string {
+	tags := make(map[string]string)
+
+	rawTags = strings.TrimPrefix(rawTags, "@")
+
+	for _, tag := range strings.Split(rawTags, ";") {
+		if !strings.Contains(tag, "=") {
+			continue
+		}
+
+		pair := strings.SplitN(tag, "=", 2)
+		key := pair[0]
+
+		var value string
+		if len(pair) == 2 {
+			value = parseIRCTagValue(pair[1])
+		}
+
+		tags[key] = value
+	}
+
+	return tags
+}
```

Comment:
This function misses some edge cases:

When a tag source doesn't have a `=`, the tag value should be `nil` IMO instead of the tag being skipped. Key-only tags are allowed: 

https://ircv3.net/specs/extensions/message-tags.html

    <tag>           ::= <key> ['=' <escaped_value>]

Also, if the `<escaped value>` is empty, the tag mapping shouldn't be `nil`, it should literally be `""` (empty string).

Reply:
I don't believe a string can be `nil`. It's "nil" value is an empty string.

I think to resolve both of these I can just remove:
```go
if !strings.Contains(tag, "=") {
	continue
}
```